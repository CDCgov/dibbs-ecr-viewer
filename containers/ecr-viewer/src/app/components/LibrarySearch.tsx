"use client";

import { Search } from "@trussworks/react-uswds";
import { useQueryParam } from "../hooks/useQueryParam";

interface LibrarySearchProps {
  className?: string;
  textBoxClassName?: string;
}

/**
 * eCR Library search bar component
 * @param props - Properties to pass into
 * @param props.className - The class name to pass into the USWDS search component
 * @param props.textBoxClassName - The class name to pass into the input props for the USWDS search component
 * @returns - Search bar component for the eCR Library
 */
const LibrarySearch = ({ className, textBoxClassName }: LibrarySearchProps) => {
  const { searchParams, updateQueryParam, pushQueryUpdate } = useQueryParam();

  return (
    <Search
      placeholder="Search by patient"
      onSubmit={(e) => {
        e.preventDefault();
        const searchTerm = (
          e.currentTarget.elements.namedItem("search-field") as HTMLInputElement
        )?.value;
        updateQueryParam("search", searchTerm);
        pushQueryUpdate();
      }}
      defaultValue={searchParams.get("search") ?? undefined}
      size="small"
      large={true}
      className={className}
      inputProps={{ className: textBoxClassName }}
    />
  );
};

export default LibrarySearch;
