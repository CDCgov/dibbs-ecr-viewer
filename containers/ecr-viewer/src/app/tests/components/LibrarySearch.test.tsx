import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import LibrarySearch from "@/app/components/LibrarySearch";

const mockPush = jest.fn();
let mockSearchParams = new URLSearchParams();
jest.mock("next/navigation", () => {
  return {
    useRouter: () => ({
      push: mockPush,
    }),
    useSearchParams: () => mockSearchParams,
    usePathname: () => "",
  };
});

describe("Library Search", () => {
  afterEach(() => {
    mockSearchParams = new URLSearchParams();
    jest.clearAllMocks();
  });

  it("should match snapshot", () => {
    const { container } = render(
      <LibrarySearch
        className="className"
        textBoxClassName="textBoxClassName"
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it("should put search term in search params", async () => {
    const user = userEvent.setup();
    render(<LibrarySearch />);
    const searchBox = screen.getByRole("searchbox");
    const searchButton = screen.getByRole("button");

    await user.type(searchBox, "Em");
    await user.click(searchButton);

    expect(mockPush).toHaveBeenCalledExactlyOnceWith("?search=Em");
  });

  it("should put set the page back to 1 after a new term is set", async () => {
    const user = userEvent.setup();
    mockSearchParams.append("page", "4");

    render(<LibrarySearch />);
    const searchBox = screen.getByRole("searchbox");
    const searchButton = screen.getByRole("button");

    await user.type(searchBox, "Em");
    await user.click(searchButton);

    expect(mockPush).toHaveBeenCalledExactlyOnceWith("?search=Em");
  });

  it("should keep the page back if the same term is searched", async () => {
    const user = userEvent.setup();
    mockSearchParams.append("page", "4");
    mockSearchParams.append("search", "Em");

    render(<LibrarySearch />);
    const searchBox = screen.getByRole("searchbox");
    const searchButton = screen.getByRole("button");

    await user.clear(searchBox);
    await user.type(searchBox, "Em");
    await user.click(searchButton);

    expect(mockPush).not.toHaveBeenCalled();
  });

  it("should put the search term in the box if already in search params", async () => {
    mockSearchParams.append("search", "blah");

    render(<LibrarySearch />);

    expect(screen.getByDisplayValue("blah")).toBeVisible();
  });
});
