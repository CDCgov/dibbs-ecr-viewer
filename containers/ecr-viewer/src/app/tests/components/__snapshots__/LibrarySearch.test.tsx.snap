// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Library Search should match snapshot 1`] = `
<div>
  <form
    class="usa-form--large usa-search className"
    data-testid="form"
    role="search"
  >
    <div
      class="textBoxClassName"
      data-testid="searchField"
    >
      <label
        class="usa-sr-only"
        data-testid="label"
        for="search-field"
      >
        Search
      </label>
      <input
        class="usa-input"
        data-testid="textInput"
        id="search-field"
        name="search"
        placeholder="Search by patient"
        type="search"
        value=""
      />
    </div>
    <div
      class="usa-search--small"
    >
      <button
        aria-label="Search"
        class="usa-button"
        data-testid="button"
        type="submit"
      >
        <svg
          aria-hidden="true"
          class="usa-icon usa-icon--size-3 usa-search__submit-icon"
          focusable="false"
          height="1em"
          name="Search"
          role="img"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15.5 14h-.79l-.28-.27A6.471 6.471 0 0 0 16 9.5 6.5 6.5 0 1 0 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
      </button>
    </div>
  </form>
</div>
`;
